File: tools/lesson_vocab.py
#!/usr/bin/env python3
"""精讲笔记词表解析与音频指纹（无第三方依赖，可供 build.py 引用）。"""

from __future__ import annotations

import hashlib
import re

DEFAULT_VOICE = "ja-JP-NanamiNeural"
DEFAULT_RATE = "-10%"

_JP_CHAR = re.compile(r"[\u3040-\u30ff\u3400-\u9fff]")
_BOLD = re.compile(r"\*\*([^*]+)\*\*")


def text_key(text: str, voice: str = DEFAULT_VOICE, rate: str = DEFAULT_RATE) -> str:
    raw = f"{text}\u0000{voice}\u0000{rate}".encode("utf-8")
    return hashlib.sha256(raw).hexdigest()[:16]


def strip_md(text: str) -> str:
    text = _BOLD.sub(r"\1", text)
    text = re.sub(r"\[([^\]]+)\]\([^)]+\)", r"\1", text)
    return text.strip()


def split_row(line: str) -> list[str]:
    line = line.strip()
    if not line.startswith("|"):
        return []
    return [p.strip() for p in line.strip("|").split("|")]


def is_sep_row(cells: list[str]) -> bool:
    if not cells:
        return False
    return all(re.fullmatch(r":?-{3,}:?", c.replace(" ", "")) for c in cells)


def has_jp(text: str) -> bool:
    return bool(_JP_CHAR.search(text))


def speak_text(word: str, reading: str) -> str:
    """决定合成 / 播放用哪段文字。"""
    word = strip_md(word)
    reading = strip_md(reading)
    if reading and has_jp(reading):
        return reading
    if word:
        return word
    return reading


def extract_speak_texts(md: str) -> list[str]:
    """抽出所有「读音」表中应合成的文本（去重保序）。"""
    lines = md.splitlines()
    texts: list[str] = []
    seen: set[str] = set()
    i = 0
    while i < len(lines):
        header = split_row(lines[i])
        if len(header) < 2:
            i += 1
            continue
        if i + 1 >= len(lines) or not is_sep_row(split_row(lines[i + 1])):
            i += 1
            continue

        reading_idx = next((j for j, h in enumerate(header) if h == "读音"), None)
        if reading_idx is None:
            i += 1
            continue
        word_idx = next(
            (j for j, h in enumerate(header) if h in ("日语", "词", "助词")),
            None,
        )
        if word_idx is None:
            word_idx = next((j for j in range(len(header)) if j != reading_idx), None)
        if word_idx is None:
            i += 1
            continue

        i += 2
        while i < len(lines):
            cells = split_row(lines[i])
            if not cells or is_sep_row(cells):
                break
            if len(cells) < max(word_idx, reading_idx) + 1:
                break
            word = cells[word_idx] if word_idx < len(cells) else ""
            reading = cells[reading_idx] if reading_idx < len(cells) else ""
            text = speak_text(word, reading)
            if text and text not in seen:
                seen.add(text)
                texts.append(text)
            i += 1
        continue
    return texts


def rewrite_vocab_links(md: str) -> str:
    """把词列表词列改成 [词](ja-word:encoded) 链接（App 端同逻辑）。"""
    lines = md.splitlines()
    out: list[str] = []
    i = 0
    while i < len(lines):
        header = split_row(lines[i])
        if (
            len(header) >= 2
            and i + 1 < len(lines)
            and is_sep_row(split_row(lines[i + 1]))
        ):
            reading_idx = next((j for j, h in enumerate(header) if h == "读音"), None)
            if reading_idx is not None:
                word_idx = next(
                    (j for j, h in enumerate(header) if h in ("日语", "词", "助词")),
                    None,
                )
                if word_idx is None:
                    word_idx = next(
                        (j for j in range(len(header)) if j != reading_idx), None
                    )
                if word_idx is not None:
                    out.append(lines[i])
                    out.append(lines[i + 1])
                    i += 2
                    while i < len(lines):
                        cells = split_row(lines[i])
                        if not cells or is_sep_row(cells):
                            break
                        if len(cells) < max(word_idx, reading_idx) + 1:
                            break
                        word = cells[word_idx]
                        reading = cells[reading_idx]
                        # 已是链接则跳过
                        if not word.startswith("["):
                            speak = speak_text(word, reading)
                            if speak:
                                from urllib.parse import quote

                                cells[word_idx] = f"[{strip_md(word)}](ja-word:{quote(speak, safe='')})"
                                lines[i] = "| " + " | ".join(cells) + " |"
                        out.append(lines[i])
                        i += 1
                    continue
        out.append(lines[i])
        i += 1
    return "\n".join(out) + ("\n" if md.endswith("\n") else "")

File: app/lib/modules/lesson/lesson_vocab.dart
// 精讲笔记 Markdown 词表：识别「读音」表并为词列注入 ja-word 链接。
// 解析规则与 tools/lesson_vocab.py 保持一致。

final _jpChar = RegExp(r'[\u3040-\u30ff\u3400-\u9fff]');
final _bold = RegExp(r'\*\*([^*]+)\*\*');
final _mdLink = RegExp(r'\[([^\]]+)\]\([^)]+\)');
final _sepCell = RegExp(r'^:?-{3,}:?$');

String stripMd(String text) {
  var t = text.replaceAllMapped(_bold, (m) => m[1]!);
  t = t.replaceAllMapped(_mdLink, (m) => m[1]!);
  return t.trim();
}

List<String> splitRow(String line) {
  final trimmed = line.trim();
  if (!trimmed.startsWith('|')) return const [];
  var inner = trimmed;
  if (inner.startsWith('|')) inner = inner.substring(1);
  if (inner.endsWith('|')) inner = inner.substring(0, inner.length - 1);
  return inner.split('|').map((c) => c.trim()).toList();
}

bool isSepRow(List<String> cells) {
  if (cells.isEmpty) return false;
  return cells.every((c) => _sepCell.hasMatch(c.replaceAll(' ', '')));
}

bool hasJp(String text) => _jpChar.hasMatch(text);

String speakText(String word, String reading) {
  final w = stripMd(word);
  final r = stripMd(reading);
  if (r.isNotEmpty && hasJp(r)) return r;
  if (w.isNotEmpty) return w;
  return r;
}

int? _wordColumnIndex(List<String> header, int readingIdx) {
  for (final name in ['日语', '词', '助词']) {
    final i = header.indexOf(name);
    if (i >= 0) return i;
  }
  for (var j = 0; j < header.length; j++) {
    if (j != readingIdx) return j;
  }
  return null;
}

/// 把带「读音」列的词表词列改成 `[词](ja-word:…)`，源文件本身不变。
String rewriteLessonVocabLinks(String md) {
  final lines = md.split('\n');
  final out = <String>[];
  var i = 0;
  while (i < lines.length) {
    final header = splitRow(lines[i]);
    if (header.length >= 2 &&
        i + 1 < lines.length &&
        isSepRow(splitRow(lines[i + 1]))) {
      final readingIdx = header.indexOf('读音');
      final wordIdx =
          readingIdx >= 0 ? _wordColumnIndex(header, readingIdx) : null;
      if (readingIdx >= 0 && wordIdx != null) {
        out.add(lines[i]);
        out.add(lines[i + 1]);
        i += 2;
        while (i < lines.length) {
          final cells = List<String>.from(splitRow(lines[i]));
          if (cells.isEmpty || isSepRow(cells)) break;
          final need = (wordIdx > readingIdx ? wordIdx : readingIdx) + 1;
          if (cells.length < need) break;
          final word = cells[wordIdx];
          final reading = cells[readingIdx];
          if (!word.startsWith('[')) {
            final speak = speakText(word, reading);
            if (speak.isNotEmpty) {
              cells[wordIdx] =
                  '[${stripMd(word)}](ja-word:${Uri.encodeComponent(speak)})';
              lines[i] = '| ${cells.join(' | ')} |';
            }
          }
          out.add(lines[i]);
          i++;
        }
        continue;
      }
    }
    out.add(lines[i]);
    i++;
  }
  final joined = out.join('\n');
  if (md.endsWith('\n') && !joined.endsWith('\n')) return '$joined\n';
  return joined;
}
